ext/src/js/modules/snapshot.js

```diff
@@ -1,190 +1,80 @@
+    br.tabs.query({
+        active: true,
+        currentWindow: true
+    }, tabs => {
+        let tab = tabs[0];
+        commChan = br.tabs.connect(
+            tab.id,
+            {name: "omnom"}
+        );
+        commChan.onMessage.addListener((msg) => {
+            const msgHandler = messageHandlers.get(msg.type);
+            if(msgHandler) {
+                msgHandler(msg);
+            } else {
+                console.log("unknown message: ", msg);
             }
+            return true;
+        });
+        if (msg) {
+            commChan.postMessage(msg);
         }
-
-    }
-    ret.html = ret.html.outerHTML;
-    return ret;
+    });
 }
 
-async function transformNode(node) {
-    if (node.nodeType !== Node.ELEMENT_NODE) {
-        return;
-    }
-    const transformFunction = nodeTransformFunctons.get(node.nodeName);
-    await rewriteAttributes(node);
-    if (transformFunction) {
-        await transformFunction(node);
-        return;
-    }
-    return;
+async function handlePongMessage(msg) {
+    numberOfPages += 1;
 }
 
-async function transformLink(node) {
-    if (node.attributes.rel && node.attributes.rel.nodeValue.trim().toLowerCase() == 'stylesheet') {
-        if (!node.attributes.href) {
-            return;
-        }
-        const index = styleIndex++;
-        const cssHref = fullURL(node.attributes.href.nodeValue);
-        const style = document.createElement('style');
-        const cssText = await downloadFile(cssHref);
-        style.innerHTML = await sanitizeCSS(cssText, cssHref);
-        styleNodes.set(index, style);
-        node.remove();
+async function handleDomDataMessage(msg) {
+    let d = new Document(msg.data.html, msg.data.url, msg.data.doctype, msg.data.attributes);
+    if (msg.data.url == getSiteUrl()) {
+        doc = d;
+    } else {
+        iframes.push(d);
     }
-    if ((node.getAttribute('rel') || '').trim().toLowerCase() == 'icon' || (node.getAttribute('rel') || '').trim().toLowerCase() == 'shortcut icon') {
-        const favicon = await downloadFile(node.href);
-        document.getElementById('favicon').value = favicon;
-        node.href = favicon;
+    if (doc && iframes.length == numberOfPages -1) {
+        doc.iframes = iframes;
+        parseDom();
     }
 }
 
-async function transformStyle(node) {
-    const innerText = await sanitizeCSS(node.innerText, getSiteUrl());
-    node.innerText = innerText;
-}
-
-async function transfromImg(node) {
-    const src = await downloadFile(node.getAttribute('src'));
-    node.src = src;
-    if (node.getAttribute('srcset')) {
-        let val = node.getAttribute('srcset');
-        let newParts = [];
-        for (let s of val.split(',')) {
-            let srcParts = s.trim().split(' ');
-            srcParts[0] = await downloadFile(srcParts[0]);
-            newParts.push(srcParts.join(' '));
-        }
-        node.setAttribute('srcset', newParts.join(', '));
+async function createSnapshot() {
```

Comment:
Imho, the create snapshot function should return with a the data needed for saveBookmark (which is a function in main.), and the communication setup, and dom request should be handled from main.js (as well as handling the saving of the bookmark)